src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/RazorPagesRazorViewEngineOptionsSetup.cs

```diff
@@ -10,29 +11,45 @@ namespace Microsoft.AspNetCore.Mvc.RazorPages.Internal
 {
     public class RazorPagesRazorViewEngineOptionsSetup : IConfigureOptions<RazorViewEngineOptions>
     {
-        private readonly IOptions<RazorPagesOptions> _pagesOptions;
+        private readonly RazorPagesOptions _pagesOptions;
 
         public RazorPagesRazorViewEngineOptionsSetup(IOptions<RazorPagesOptions> pagesOptions)
         {
-            _pagesOptions = pagesOptions;
+            _pagesOptions = pagesOptions?.Value ?? throw new ArgumentNullException(nameof(pagesOptions));
         }
 
         public void Configure(RazorViewEngineOptions options)
         {
-            Debug.Assert(_pagesOptions.Value.RootDirectory.Length > 0);
-
-            if (_pagesOptions.Value.RootDirectory == "/")
-            {
-                options.PageViewLocationFormats.Add("/{1}/{0}" + RazorViewEngine.ViewExtension);
-            }
-            else
+            if (options == null)
             {
-                options.PageViewLocationFormats.Add(_pagesOptions.Value.RootDirectory + "/{1}/{0}" + RazorViewEngine.ViewExtension);
+                throw new ArgumentNullException(nameof(options));
             }
 
+            var rootDirectory = _pagesOptions.RootDirectory;
+            Debug.Assert(!string.IsNullOrEmpty(rootDirectory));
+            var defaultPageSearchPath = CombinePath(rootDirectory, "{1}/{0}");
+            options.PageViewLocationFormats.Add(defaultPageSearchPath);
+
+            // /Pages/Shared/{0}.cshtml
+            var pagesSharedDirectory = CombinePath(rootDirectory, "Shared/{0}");
+            options.PageViewLocationFormats.Add(pagesSharedDirectory);
+
             options.PageViewLocationFormats.Add("/Views/Shared/{0}" + RazorViewEngine.ViewExtension);
 
+            options.ViewLocationFormats.Add(pagesSharedDirectory);
+            options.AreaViewLocationFormats.Add(pagesSharedDirectory);
+
             options.ViewLocationExpanders.Add(new PageViewLocationExpander());
         }
+
+        private static string CombinePath(string path1, string path2)
+        {
+            if (path1.EndsWith("/", StringComparison.Ordinal))
+            {
+                return path1 + path2 + RazorViewEngine.ViewExtension;
+            }
+
+            return path1 + "/" + path2 + RazorViewEngine.ViewExtension;
```

Comment:
since path1 is always `_pagesOptions.RootDirectory` do we guarantee the absence of a trailing slash? or do we allow that?

Reply:
That's what the check in line #47 does, no?